grovedb/src/lib.rs

```diff
@@ -278,6 +288,86 @@ impl GroveDb {
         Ok(proof_result)
     }
 
+    pub fn verify_proof(
+        path: &[&[u8]],
+        proofs: &mut Vec<Vec<u8>>, // Generic into_iterator (trait) u8
+        expected_root_hash: [u8; 32],
+    ) -> Result<Map, Error> {
+        if proofs.len() < 2 {
+            return Err(Error::InvalidProof("Proof length should be 2 or more"));
+        }
+
+        if proofs.len() - 2 != path.len() {
+            return Err(Error::InvalidProof(
+                "Proof length should be two greater than path",
+            ));
+        }
+
+        let root_leaf_keys: HashMap<Vec<u8>, usize> =
+            bincode::deserialize(&proofs.pop().unwrap()[..])?;
+
+        let mut proof_iterator = proofs.iter();
+        let reverse_path_iterator = path.iter().rev();
+
+        let leaf_proof = proof_iterator
+            .next()
+            .expect("Constraint checks above enforces leaf proof must exist");
+
+        let (mut last_root_hash, leaf_result_map) = match execute_proof(&leaf_proof[..]) {
+            Ok(result) => Ok(result),
+            Err(e) => Err(Error::InvalidProof("Invalid proof element")),
+        }?;
+
+        let mut proof_path_zip = proof_iterator.zip(reverse_path_iterator).peekable();
+
+        while let Some((proof, key)) = proof_path_zip.next() {
```

Comment:
`while let Some(x) = iterator.next()` looks like `for x in iterator` for me, what do you think?

Reply:
oh it's because of `peek` I see, nvm then

Reply:
Yeah